### PR TITLE
docs: add proof doc

### PR DESCRIPTION
**Pull request by tnachen**

### CODE REVIEW

File: docs/proof/proof.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+One can verify a root hash and compares with the root hash that is computed from the RangeProof is equal.
+As long as the tree content and the hashing mechanism remains the same, the root hash over a merkle tree should always be the same.
+This is used to verify that a `RangeProof` is valid, since the root hash computed from the proof and tree should be identical.
+  
+## VerifyItem
+
+Once a `RangeProof` is verified it can be used to prove that a key/value within the range exists in this
+`RangeProof`. Since all the keys as part of the `RangeProof` is stored in Leaves, verifying a particular key and value 
+is simply verifying one of the leaf node's key matches the key and its value's hash matches the valuehash stored.
```
Comment (tessr) [inline docs/proof/proof.md]:
```suggestion
is done by simply verifying that one of the leaf node's key matches the key and its value's hash matches the valuehash stored.
```

Comment (tessr) [reply docs/proof/proof.md]: (Stored where?)

File: docs/proof/proof.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+## VerifyAbsence
+
+Similarly to VerifiyItem, once a `RangeProof` is verified it can be used to prove that a key/value within the range
+doesn't exist. Proving that a key doesn't exist from a `RangeProof` is accomplished by verifying that the key
+doesn't exist in the Leaves.
```
Comment (tessr) [inline docs/proof/proof.md]: Would be nice to give more information on how this actually works. (I assume we don't have to iterate over all the leaves to verify this, for example.)

File: proof_range.go [hunk 3]
```diff
@@ -396,10 +395,10 @@ func (t *ImmutableTree) getRangeProof(keyStart, keyEnd []byte, limit int) (proof
 			}
 
 			if node.height == 0 {
-				// Leaf node.
-				// Append inners to innersq.
-				innersq = append(innersq, inners)
-				inners = PathToLeaf(nil)
+				// Append inners (PathToLeaf) that we tracked so far to get to this leaf node to innersq.
```
Comment (tessr) [inline proof_range.go]: Looks like the variables referenced here got renamed

File: go.sum [hunk 4]
```diff
@@ -182,8 +179,6 @@ github.com/tecbot/gorocksdb v0.0.0-20191017175515-d217d93fd4c5 h1:gVwAW5OwaZlDB5
 github.com/tecbot/gorocksdb v0.0.0-20191017175515-d217d93fd4c5/go.mod h1:ahpPrc7HpcfEWDQRZEmnXMzHY03mLDYMCxeDzy46i+8=
 github.com/tendermint/go-amino v0.14.1 h1:o2WudxNfdLNBwMyl2dqOJxiro5rfrEaU0Ugs6offJMk=
 github.com/tendermint/go-amino v0.14.1/go.mod h1:i/UKE5Uocn+argJJBb12qTZsCDBcAYMbR92AaJVmKso=
-github.com/tendermint/tendermint v0.33.0 h1:TW1g9sQs3YSqKM8o1+opL3/VmBy4Ke/VKV9MxYpqNbI=
-github.com/tendermint/tendermint v0.33.0/go.mod h1:s5UoymnPIY+GcA3mMte4P9gpMP8vS7UH7HBXikT1pHI=
```
Comment (tessr) [inline go.sum]: why all the mod changes?

File: proof_range.go [hunk 5]
```diff
@@ -425,14 +424,19 @@ func (t *ImmutableTree) getRangeProof(keyStart, keyEnd []byte, limit int) (proof
 				}
 			} else {
 				// Inner node.
-				if pathCount >= 0 {
-					// Skip redundant path items.
-				} else {
-					inners = append(inners, ProofInnerNode{
+				if pathCount < 0 {
+					// Only store in inners if the node is not stored in PathToLeaf already.
```
Comment (tessr) [inline proof_range.go]: Another renamed variable.

**Comment by erikgrinaker** [comment]
I'll pick this up and expand it a bit.

**Comment by erikgrinaker** [comment]
Code changes have been split out to #248.

**Comment by erikgrinaker** [comment]
@tessr I expanded quite a bit on Tim's original draft, rewriting most of it - let me know what you think.

File: docs/proof/proof.md [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+err = proof.VerifyItem([]byte("b"), []byte{2})
+if err != nil {
+    log.Printf("Failed to prove b=2 with range: %v", err)
+}
+
+// Since "ab" is ordered after "a" but before "b", we can prove that it
+// is not in the range and therefore not in the tree at all
+err = proof.VerifyAbsence([]byte("ab"))
+if err != nil {
+    log.Printf("Failed to prove absence of ab: %v", err)
```
Comment (tessr) [inline docs/proof/proof.md:249]: This is pretty nitpicky, but I think these examples would be a little bit easier to read if it were clearer when we were expecting an error vs. not--every function call is followed by an error check regardless of whether we expect the error or not.

File: docs/proof/proof.md [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+With this data, we are able to compute the hashes for all nodes up to and including the root,
+and can compare this root hash with the root hash of the IAVL tree - if they match, we can be
+reasonably certain that the value is correct. This data is therefore considered a _proof_ for the
+value. Notice how we don't need to include any data from e.g. the `e`-branch of the tree at all,
+only the hash - as the tree grows in size, these savings become very significant, requiring only
+`log₂(n)` hashes for a tree of `n` keys.
+
+However, this still introduces quite a bit of overhead. Since we usually want to fetch several
+values from the tree and verify them, it is often useful to fetch a _range proof_, which contains
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply docs/proof/proof.md:57]: I'm not overly familiar with this domain, but from what I understand the idea is that a range proof for the range `c-h` exactly proves all values that are and are not present in the range from c (inclusive) to h (exclusive). This is regardless of whether any keys exist at all in this range, including the endpoints `c` and `h`. I can try to clarify this a bit.

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply docs/proof/proof.md]: Tried explaining it a bit better now, did it help?

File: docs/proof/proof.md (hunk 6 above)
Comment (erikgrinaker) [reply docs/proof/proof.md]: Changed it to printing the result, with a comment of the output - is that better?

**Pull request merged**